connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DeadLetterQueueReporter.java

```diff
@@ -118,11 +140,73 @@ public void report(ProcessingContext context) {
         }
 
+        if (connConfig.isDlqContextHeadersEnabled()) {
+            populateContextHeaders(producerRecord, context);
+        }
+
         this.kafkaProducer.send(producerRecord, (metadata, exception) -> {
             if (exception != null) {
                 log.error("Could not produce message to dead letter queue. topic=" + dlqTopicName, exception);
                 errorHandlingMetrics.recordDeadLetterQueueProduceFailed();
             }
         });
     }
+
+    // Visible for testing
+    void populateContextHeaders(ProducerRecord<byte[], byte[]> producerRecord, ProcessingContext context) {
+        String topic = "";
+        int partition = -1;
+        long offset = -1;
+        if (context.consumerRecord() != null) {
+            topic = context.consumerRecord().topic();
+            partition = context.consumerRecord().partition();
+            offset = context.consumerRecord().offset();
+        }
+
+        add(producerRecord.headers(), prefix(ERROR_HEADER_ORIG_TOPIC), topic);
+        add(producerRecord.headers(), prefix(ERROR_HEADER_ORIG_PARTITION), String.valueOf(partition));
+        add(producerRecord.headers(), prefix(ERROR_HEADER_ORIG_OFFSET), String.valueOf(offset));
+        add(producerRecord.headers(), prefix(ERROR_HEADER_CONNECTOR_NAME), connectorTaskId.connector());
+        add(producerRecord.headers(), prefix(ERROR_HEADER_TASK_ID), String.valueOf(connectorTaskId.task()));
+        add(producerRecord.headers(), prefix(ERROR_HEADER_STAGE), context.stage().name());
+        add(producerRecord.headers(), prefix(ERROR_HEADER_EXECUTING_CLASS), context.executingClass().getName());
+        if (context.error() != null) {
+            add(producerRecord.headers(), prefix(ERROR_HEADER_EXECPTION), context.error().getClass().getName());
+            add(producerRecord.headers(), prefix(ERROR_HEADER_EXECPTION_MESSAGE), context.error().getMessage());
+            byte[] trace;
+            if ((trace = stacktrace(context.error())) != null) {
+                add(producerRecord.headers(), prefix(ERROR_HEADER_EXECPTION_STACK_TRACE), trace);
+            }
+        }
+    }
+
+    private byte[] stacktrace(Throwable error) {
+        ByteArrayOutputStream bos = new ByteArrayOutputStream();
+        try {
+            PrintStream stream = new PrintStream(bos, true, "UTF-8");
+            error.printStackTrace(stream);
+            bos.close();
+            return bos.toByteArray();
+        } catch (IOException e) {
+            log.error("Could not serialize stacktrace.", e);
+        }
+        return null;
+    }
+
+    private void add(Headers headers, String key, String value) {
+        add(headers, key, value.getBytes(Charset.forName("UTF-8")));
+    }
+
+    private void add(Headers headers, String key, byte[] value) {
+        if (headers.lastHeader(key) == null) {
+            headers.add(key, value);
+        } else {
+            log.error("Header already contains the '" + key + "' key.");
```

Comment:
Seems like this might occur only in a programming error, which we could test for in a unit test.